lib/actions/api.js

```diff
@@ -17,32 +19,113 @@ export const nonRealtimeRoutingResponse = createAction('NON_REALTIME_ROUTING_RES
+  let activeItinerary = otpState.currentQuery.routingType === 'ITINERARY' ? 0 : null
+  // We cannot use window.history.state here to check for the active
+  // itinerary param because it is unreliable in some states (e.g.,
+  // when the print layout component first loads).
+  const urlParams = getUrlParams()
+  if (
+    (!otpState.searches || Object.keys(otpState.searches).length === 0) &&
+    urlParams.ui_activeItinerary
+  ) {
+    activeItinerary = +urlParams.ui_activeItinerary
+  }
+  return activeItinerary
+}
 
-export function routingQuery () {
+/**
+ * Send a routing query to the OTP backend.
+ *
+ * NOTE: We need a random ID so that when a user reloads the page (clearing the
+ * state), performs searches, and presses back to load previous searches
+ * that are no longer contained in the state we don't confuse the search IDs
+ * with search IDs from the new session. If we were to use sequential numbers
+ * as IDs, we would run into this problem.
+ */
+export function routingQuery (searchId = null) {
   return async function (dispatch, getState) {
     const otpState = getState().otp
+    const isNewSearch = !searchId
+    if (isNewSearch) searchId = randId()
     const routingType = otpState.currentQuery.routingType
-    const searchId = ++lastSearchId
-
-    if (!queryIsValid(otpState)) return
-    dispatch(routingRequest({ routingType, searchId }))
+    // Don't permit a routing query if the query is invalid
+    if (!queryIsValid(otpState)) {
+      console.warn('Query is invalid. Aborting routing query', otpState.currentQuery)
+      return
+    }
+    const activeItinerary = getActiveItinerary(otpState)
+    dispatch(routingRequest({ activeItinerary, routingType, searchId }))
 
     // fetch a realtime route
-    fetch(constructRoutingQuery(otpState))
+    const query = constructRoutingQuery(otpState)
+    fetch(query)
       .then(getJsonAndCheckResponse)
       .then(json => {
         dispatch(routingResponse({ response: json, searchId }))
+        // If tracking is enabled, store locations and search after successful
+        // search is completed.
+        // TODO recent searches
+        if (otpState.user.trackRecent) {
+          const { from, to } = otpState.currentQuery
+          if (!isStoredPlace(from)) {
+            dispatch(rememberPlace({ type: 'recent', location: formatRecentPlace(from) }))
+          }
+          if (!isStoredPlace(to)) {
+            dispatch(rememberPlace({ type: 'recent', location: formatRecentPlace(to) }))
+          }
+          dispatch(rememberSearch(formatRecentSearch(query, otpState)))
+        }
       })
       .catch(error => {
         dispatch(routingError({ error, searchId }))
       })
-
+    // Update OTP URL params if a new search. In other words, if we're
+    // performing a search based on query params taken from the URL after a back
+    // button press, we don't need to update the OTP URL.
+    // TODO: Should we be replacing the URL params here?
```

Comment:
Seems like it'd be a good idea to replace the url params whenever a state change occurs that is reproducible based on the given url params.

Reply:
I need to elaborate on this TODO. I was running into issues where if the app was re-running an old search found in the URL params and it pushed a new path with the query params it would funk with the history.  The TODO should say `TODO: For old searches that we are re-running, should we be **replacing** the URL params here (instead of **pushing** a new path to history like what currently happens in updateOtpUrlParams)? That way we could ensure that the path absolutely accurately reflects the app state.`

Reply:
Ah, I see. Let's await user feedback and stuff and keep that expanded TODO comment.